server/src/main/java/org/apache/iotdb/db/mpp/memory/MemoryPool.java

```diff
@@ -79,11 +147,35 @@ public synchronized void free(String queryId, long bytes) {
     } else {
       queryMemoryReservations.put(queryId, queryReservedBytes);
     }
-
     reservedBytes -= bytes;
+
+    if (memoryReservationFutures.isEmpty()) {
+      return;
+    }
+    Iterator<MemoryReservationFuture<Void>> iterator = memoryReservationFutures.iterator();
+    while (iterator.hasNext()) {
+      MemoryReservationFuture<Void> future = iterator.next();
+
+      if (future.isCancelled()) {
+        iterator.remove();
+        continue;
+      }
+
+      long bytesToReserve = future.getBytes();
+      if (maxBytes - reservedBytes < bytesToReserve
+          || maxBytesPerQuery - queryMemoryReservations.getOrDefault(future.getQueryId(), 0L)
+              < bytesToReserve) {
+        return;
```

Comment:
We need to separate these two conditions here. If the maxBytes is not hit but the maxBytesPerQuery is hit, we need to continue to judge whether next future is satisfied or not because next future may belong to another query.

```
if (maxBytes - reservedBytes < bytesToReserve) {
   return;
}
if (maxBytesPerQuery - queryMemoryReservations.getOrDefault(future.getQueryId(), 0L)
              < bytesToReserve) {
   continue;
}
```

Reply:
Fixed.